src/pages/resultsView/ResultsViewPageStore.ts

```diff
@@ -175,6 +176,55 @@ export function countAlterationOccurences(samplesByCancerType: {[cancerType: str
 
 }
 
+export function extendSamplesWithCancerType(samples:Sample[], clinicalDataForSamples:ClinicalData[], studies:CancerStudy[]){
+
+    const clinicalDataGroupedBySampleId = _.groupBy(clinicalDataForSamples, (clinicalData:ClinicalData)=>clinicalData.uniqueSampleKey);
+    // note that this table is actually mutating underlying sample.  it's not worth it to clone samples just
+    // for purity
+    const extendedSamples = samples.map((sample: ExtendedSample)=>{
+        const clinicalData = clinicalDataGroupedBySampleId[sample.uniqueSampleKey];
+        if (clinicalData) {
+            clinicalData.forEach((clinicalDatum:ClinicalData)=>{
+                switch (clinicalDatum.clinicalAttributeId) {
+                    case 'CANCER_TYPE_DETAILED':
+                        sample.cancerTypeDetailed = clinicalDatum.value;
+                        break;
+                    case 'CANCER_TYPE':
+                        sample.cancerType = clinicalDatum.value;
+                        break;
+                    default:
+                        break;
+                }
+            });
+        }
+        return sample;
+    });
+
+    //make a map by studyId for easy access in following loop
+    const studyMap = _.keyBy(studies,(study:CancerStudy)=>study.studyId);
+
+    // now we need to fix any samples which do not have both cancerType and cancerTypeDetailed
+    extendedSamples.forEach((sample:ExtendedSample)=>{
+        //if we have no cancer subtype, then make the subtype the parent type
+        if (!sample.cancerType) {
+            // we need to fall back to studies cancerType
+            const study = studyMap[sample.studyId];
+            if (study) {
+                sample.cancerType = study.cancerType.name;
+            } else {
+                sample.cancerType = "Unknown";
+            }
+        }
+        if (sample.cancerType && !sample.cancerTypeDetailed) {
+            sample.cancerTypeDetailed = sample.cancerType;
+        }
+    });
+
+    return extendedSamples;
+
+}
```

Comment:
maybe for clarity (I dont know if necessary) but we could have
```
export interface ExtendedSample extends Sample {
    clinCancerType: string;
    clinCancerTypeDetailed: string;
    studyCancerType: string;
    cancerType: string; // derived based on which of the above three are available
}
```

This may be useful?

Reply:
I see.  Yeah.  I like it.